dali/operators/math/expressions/broadcasting.cc

```diff
@@ -0,0 +1,242 @@
+          if (res > 1)
+            DALI_FAIL(make_string("Can't broadcast extents ", extent, " and ", res));
+          res = extent;
+        }
+      }
+    }
+  }
+}
+
+template <typename Shape>
+bool CanBroadcastImpl(span<Shape> shapes) {
+  if (shapes.size() < 2) {
+    return true;
+  }
+  int ndim = BroadcastNdim(shapes);
+  for (int d = 0; d < ndim; d++) {
+    int64_t val = 1;
+    for (int i = 0; i < shapes.size(); i++) {
+      auto &sh = shapes[i];
+      auto extent = d < sh.size() ? sh[sh.size() - 1 - d] : 1;
+      if (extent > 1 && extent != val) {
+        if (val > 1)
+          return false;
+        val = extent;
+      }
+    }
+  }
+  return true;
+}
+
+bool CanBroadcast(span<const TensorShape<>> shapes) {
+  return CanBroadcastImpl(shapes);
+}
+
+bool CanBroadcast(span<const TensorListShape<>> shapes) {
+  if (shapes.size() < 2) {
+    return true;
+  }
+  int ndim = BroadcastNdim(shapes);
+  int nsamples = shapes[0].num_samples();
+  for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+    for (int d = 0; d < ndim; d++) {
+      int64_t val = 1;
+      for (int i = 0; i < shapes.size(); i++) {
+        const auto &sh = shapes[i].tensor_shape_span(sample_idx);
+        auto extent = d < sh.size() ? sh[sh.size() - 1 - d] : 1;
+        if (extent > 1 && extent != val) {
+          if (val > 1)
+            return false;
+          val = extent;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+template <typename Shape>
+bool NeedBroadcastImpl(span<Shape> shapes) {
+  if (shapes.size() < 2)
+    return false;
+  auto *prev_sh = &shapes[0];
+  for (int i = 1; i < shapes.size(); i++) {
+    if (IsScalarLike(*prev_sh)) {
+      prev_sh = &shapes[i];
+    } else if (!IsScalarLike(shapes[i]) && *prev_sh != shapes[i]) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool NeedBroadcasting(span<const TensorShape<>> shapes) {
+  return NeedBroadcastImpl(shapes);
+}
+
+bool NeedBroadcasting(span<const TensorListShape<>> shapes) {
+  return NeedBroadcastImpl(shapes);
+}
+
+TensorShape<> StridesForBroadcasting(const TensorShape<> &out_sh, const TensorShape<> &in_sh,
+                                     const TensorShape<> &in_strides) {
+  TensorShape<> strides;
+  assert(in_sh.size() == in_strides.size());
+  assert(in_sh.size() <= out_sh.size());
+  int out_ndim = out_sh.size();
+  strides.shape.resize(out_ndim, 0);
+  for (int i = (out_ndim - in_sh.size()); i < out_ndim; i++) {
+    assert(in_sh[i] == out_sh[i] || in_sh[i] == 1);
+    if (in_sh[i] == out_sh[i])
+      strides[i] = in_strides[i];
+    else
+      strides[i] = 0;
+  }
+  return strides;
+}
+
+void ExpandToNDims(TensorShape<> &sh, int ndim) {
+  assert(sh.size() <= ndim);
+  if (sh.size() == ndim)
+    return;
+  TensorShape<> sh2;
+  sh2.shape.resize(ndim);
+  int i = 0;
+  for (; i < (ndim - sh.size()); i++)
+    sh2[i] = 1;
+  for (int j = 0; j < sh.size(); j++, i++)
+    sh2[i] = sh[j];
+  std::swap(sh, sh2);
```

Comment:
Don't we have shape concatenation?
`shape_cat(<some_helper_to_generate {1, ... 1} of given ndim>,  sh);`